apps: fixed rpc_demo for master and slave Linux machine

viswaharsha opened the pull request:
**Modified for Remote procedure call between master and slave Linux systems** 

- added definition for `SCAN_SYSCALL_ID`
- added `_scanf` function to perform input operation in master implementing `rpmsg_send `function
- modified `xil_printf` to `generic printf` to be compatible with Linux machine
- modified `open`, `read`, `write`, `close` local file system calls to `_open`, `_read`, `_write`, `_close` rpmsg calls
- modified `scanf`to `_scanf`to get rpmsg based input through master
- added `handle_scanf`to perform local `scanf`function in master implementing `rpmsg_send`call
- added case in `handle_rpc` to redirect to `handle_scanf` function
- added `rpc_demo` to the list of `app_list`

Signed-off-by: ViswaHarsha C <cviswa.harsha_ext@ltts.com>

viswaharsha commented:
Thanks for the comments. However, I feel to understand Openamp working and rpmsg/rpc_demo, it would be better if it could work solely on a Linux System. Probably, I will add a macro as a LINUX_DEMO which could help people like me who before porting it to a board can work with the code and run the application on a Linux system.

I hope this helps. Let me send you another commit with LINUX_DEMO macro. Let me know.

arnopo commented:
> Thanks for the comments. However, I feel to understand Openamp working and rpmsg/rpc_demo, it would be better if it could work solely on a Linux System. Probably, I will add a macro as a LINUX_DEMO which could help people like me who before porting it to a board can work with the code and run the application on a Linux system.
> 
> I hope this helps. Let me send you another commit with LINUX_DEMO macro. Let me know.

Sure, your contribution is welcome. Just be sure to not break the legacy.

An alternative might be to develop a new rpc demo. This would be a copy/paste in terms of the application's behavior, except that it would not rely on syscalls but on a more generic rpc proxy service (with one client and one server).  

The rpc communication could be based on a generic structure send on rpmsg, something like
```
struct rpc_rpmsg {
  char rpc_name [20];
  char rpc_param[0];
}
```
On server side application would register a table containing a list of supported rpc_name. On client side, a stub function would packs the function name and parameters in the rpc_rpmsg struct and send the rpmsg.

viswaharsha commented:
Thanks for the comments. As suggested, I will create a separate file. I understand that in the client side you want me to change it to a generic structure. I totally agree that it can be a generic struct. For more clarity can you confirm my understanding that you mean that I change the _scanf to take this generic structure as parameters or changes needed for other functions like read, write also? 

Also I am unable to understand why a change in server side is required as there is a handle_scanf aldready?  Can you elaborate a little bit more.

Can we have a short call with you to understand better.

arnopo commented:
> Thanks for the comments. As suggested, I will create a separate file. I understand that in the client side you want me to change it to a generic structure. I totally agree that it can be a generic struct. For more clarity can you confirm my understanding that you mean that I change the _scanf to take this generic structure as parameters or changes needed for other functions like read, write also?

What i suggest in https://github.com/OpenAMP/open-amp/pull/290#issuecomment-853678589 is:
- Don't use open-amp/lib/proxy/rpmsg_retarget.c if it you can not use it. Instead, create a new rpmsg service as suggested in 
   In this case put the service in  open-amp/lib/service/rpmsg/rpc/rpmsg_rpc_client.c and open-amp/lib/service/rpmsg/rpc/rpmsg_rpc_server.c
   => new folder containing generic services over rpmsg ( platform agnostic).
 
- create a new application exemple in  apps/examples/rpc/ that relies on the new rpmsg service instead of rpmsg_retarget.
  This application could be a copy past of the rpc_demo in its way of working with adaptation to use the new rpc service for the open/close/read/write/printfand scanf.

> 
> Also I am unable to understand why a change in server side is required as there is a handle_scanf aldready? Can you elaborate a little bit more.

The handle_scanf that you introduced in your patch seems to me ok. I guess you also think you have a handle_printf...
My concerns is the[ _scanf](https://github.com/OpenAMP/open-amp/pull/290/commits/79a530f5bca7c2b64d84aaab93039548c3b6e79b#diff-acf70c850987229010bf756028d506ae05c2d2864f179acf9c37373f424b203b) that should not part of the syscalls

> 
> Can we have a short call with you to understand better.

The biweekly meeting is a good place for this kind of discussion.  It would also be a good opportunity to share with others the rationale for using rpmsg for interprocessor communication.

arnopo commented:
hi @viswaharsha

Please could you restructure your patchset to make it simpler to review. For this don't hesitate to overwrite your existing patches instead of adding some revert. To ease the review it is also nice to split the patches per feature. For instance for your patchset, one commit related to the lib/service/rpmsg/rpc/rpmsg_rpc_client , one related to the application.

To do this just use  git push -f to force the update on your branch on your github, the PR will be automatically updated

Thanks

viswaharsha commented:
**Remote procedure call between Linux master and slave machines**

**Services**
- added `rpmsg_open`, `rpmsg_read`, `rpmsg_write`, `rpmsg_close` for file operations
- added definition for `SCAN_SYSCALL_ID`
- added `rpmsg_scanf` for input operation

viswaharsha commented:
**Remote procedure call between Linux master and slave machines**

**Services**

- added rpmsg_open, rpmsg_read, rpmsg_write, rpmsg_close for file operations
- added definition for SCAN_SYSCALL_ID
- added rpmsg_scanf for input operation

viswaharsha commented:
**Remote procedure call between Linux master and slave machines**

 

**Application**
- added `linux_rpc_demod.c` for linux master
- added `linux_rpc_demo.c` for linux slave

viswaharsha reviewed (commented):
**Remote procedure call between Linux master and slave machines**

**Services**	

- added `rpmsg_open`, `rpmsg_read`, `rpmsg_write`, `rpmsg_close` for file operations
- added definition for `SCAN_SYSCALL_ID`
- added `rpmsg_scanf` for input operations

viswaharsha reviewed (commented):
**Remote procedure call between Linux master and slave machines**

 

**Application**
- added `linux_rpc_demod.c` for linux master
- added `linux_rpc_demo.c` for linux slave

arnopo reviewed (changes requested):
rpmsg_rpc_client is not a generic rpc rpmsg service but a part of the application.
rpmsg_retarget is a rpc for syscall service,  so the service API is at syscall API level.
You need more rpc services, so your rpmsg service APi should be only a rpc API :
* rpc init
* rpc_send
* Callback mechanism to get rpc result

viswaharsha commented:
Hi @arnopo , below are the added changes to make a generic rpmsg rpc service. Could you have a look at it and confirm if these are the changes requested, so that I can send the PR for review which is currently under testing.

**Client Service**(`service/rpmsg/rpc/pmsg_rpc_client.c`)

- added `linux_rpmsg_rpc_init` to load the rpc
- added `rpmsg_rpc_client_send` to take in function id and request parameters, packed as character array for `rpmsg_send` with no poll service, exactly as the code snippet shared.
- added `linux_rpmsg_endpoint_client_cb` to unpack the character array to get function id and parameters

**Server Service**(`service/rpmsg/rpc/rpmsg_rpc_server.c`)

- added `rpc_app_callback` to do a application function call back based on function id
- added `linux_rpmsg_endpoint_server_cb` to unpack character array to get function id  and parameters and do `rpc_app_callback`
- added `rpmsg_rpc_server_send` to take in function id and response parameters, packed as character array for `rpmsg_send`


**Client Server Header**(`rpmsg_rpc_client_server.h`)

- added request and response structures declarations for `open`, `read`, `write`, `close`, `input`
- added all function declarations for client and server service.
- declared definitions for function ids

**Slave application**(`linux_rpc_demo.c`)
- added definitions for `rpmsg_open`, `rpmsg_close`, `rpmsg_read`, `rpmsg_write`, `rpmsg_input` to send the request parameters to `rpmsg_rpc_client_send`, exactly as the code snippet shared.

**Master application**(`linux_rpc_demod.c`)

- added definitions for `rpmsg_handle_open`, `rpmsg_handle_close`, `rpmsg_handle_read`,
`rpmsg_handle_write`, `rpmsg_handle_input` to perform the function operation from received request and pack the response for `rpmg_rpc_server_send`

arnopo commented:
> Hi @arnopo , below are the added changes to make a generic rpmsg rpc service. Could you have a look at it and confirm if these are the changes requested, so that I can send the PR for review which is currently under testing.

This is seems to match which what i have in mind. pleasefind few remarks below:

> 
> **Client Service**(`service/rpmsg/rpc/pmsg_rpc_client.c`)
> 
> * added `linux_rpmsg_rpc_init` to load the rpc
> * added `rpmsg_rpc_client_send` to take in function id and request parameters, packed as character array for `rpmsg_send` with no poll service, exactly as the code snippet shared.
> * added `linux_rpmsg_endpoint_client_cb` to unpack the character array to get function id and parameters
> 
> **Server Service**(`service/rpmsg/rpc/rpmsg_rpc_server.c`)
> 
> * added `rpc_app_callback` to do a application function call back based on function id
> * added `linux_rpmsg_endpoint_server_cb` to unpack character array to get function id  and parameters and do `rpc_app_callback`
> * added `rpmsg_rpc_server_send` to take in function id and response parameters, packed as character array for `rpmsg_send`

Seems that a linux_rpmsg_rpc_server_init API is missing here. This API would be sused by application to transmit a table of supported service containing  function ID and their associated  calback) 

> 
> **Client Server Header**(`rpmsg_rpc_client_server.h`)
> 
> * added request and response structures declarations for `open`, `read`, `write`, `close`, `input`
> * added all function declarations for client and server service.
> * declared definitions for function ids
> 

rpmsg_rpc_client_server.h should be part of the applications right? The rpc rpmsg service don't have to know feature associated rpc functions ( such as open, read, write...).

> **Slave application**(`linux_rpc_demo.c`)

Is it appliacble only for linux or could it be compiled for another platform?

> 
> * added definitions for `rpmsg_open`, `rpmsg_close`, `rpmsg_read`, `rpmsg_write`, `rpmsg_input` to send the request parameters to `rpmsg_rpc_client_send`, exactly as the code snippet shared.
> 
> **Master application**(`linux_rpc_demod.c`)
> 
> * added definitions for `rpmsg_handle_open`, `rpmsg_handle_close`, `rpmsg_handle_read`,
>   `rpmsg_handle_write`, `rpmsg_handle_input` to perform the function operation from received request and pack the response for `rpmg_rpc_server_send`

Thanks
Arnaud

viswaharsha commented:
Hi @arnopo ,

Thanks, I had named it as rpc_app_callback, based on you comment as it makes sense will rename it to `linux_rpmsg_rpc_server_init`
> Seems that a linux_rpmsg_rpc_server_init API is missing here. This API would be sused by application to transmit a table of supported service containing function ID and their associated calback)
> 
> > **Client Server Header**(`rpmsg_rpc_client_server.h`)
> > 
> > * added request and response structures declarations for `open`, `read`, `write`, `close`, `input`
> > * added all function declarations for client and server service.
> > * declared definitions for function ids
> 
Yes you are right, it is part of application. Will remote it from the header and do the needful.
> rpmsg_rpc_client_server.h should be part of the applications right? The rpc rpmsg service don't have to know feature associated rpc functions ( such as open, read, write...).
> 
> > **Slave application**(`linux_rpc_demo.c`)
> 
Currently I have focused on only Linux platforms. However, I feel it could be extended for other platforms in future. Restricting my commit currently only for Linux (as I have tested on only Linux) .
> Is it appliacble only for linux or could it be compiled for another platform?
> 
> > * added definitions for `rpmsg_open`, `rpmsg_close`, `rpmsg_read`, `rpmsg_write`, `rpmsg_input` to send the request parameters to `rpmsg_rpc_client_send`, exactly as the code snippet shared.
> > 
> > **Master application**(`linux_rpc_demod.c`)
> > 
> > * added definitions for `rpmsg_handle_open`, `rpmsg_handle_close`, `rpmsg_handle_read`,
> >   `rpmsg_handle_write`, `rpmsg_handle_input` to perform the function operation from received request and pack the response for `rpmg_rpc_server_send`
> 

Thanks
Viswa Harsha C

arnopo commented:
> > associated rpc functions ( such as open, read, write...).
> > > **Slave application**(`linux_rpc_demo.c`)
> 
> Currently I have focused on only Linux platforms. However, I feel it could be extended for other platforms in future. Restricting my commit currently only for Linux (as I have tested on only Linux) .
No need to test application on all platforms, if  no code specific to linux application, just supress "linux" if application file names.

viswaharsha commented:
Hi @arnopo , added the following changes to provide a generic rpmsg rpc service as requested. 

**Client Service(service/rpmsg/rpc/pmsg_rpc_client.c)**

- added linux_rpmsg_rpc_init to load the rpc
- added rpmsg_rpc_client_send to take in function id and request parameters, packed as character array for rpmsg_send with no poll service, exactly as the code snippet shared.
- added linux_rpmsg_endpoint_client_cb to unpack the character array to get function id and parameters

**Server Service(service/rpmsg/rpc/rpmsg_rpc_server.c)**

- added linux_rpmsg_rpc_server_init to do a application function call back based on function id
- added linux_rpmsg_endpoint_server_cb to unpack character array to get function id and parameters and do rpc_app_callback
- added rpmsg_rpc_server_send to take in function id and response parameters, packed as character array for rpmsg_send
- added definitions for rpmsg_handle_open, rpmsg_handle_close, rpmsg_handle_read, rpmsg_handle_write, rpmsg_handle_input to perform the function operation from received request and pack the response for rpmg_rpc_server_send

**Client Server Header(rpmsg_rpc_client_server.h)**

- added request and response structures declarations for open, read, write, close, input
- added all function declarations for client and server service.
- declared definitions for function ids

**Slave application(linux_rpc_demo.c)**

- added definitions for rpmsg_open, rpmsg_close, rpmsg_read, rpmsg_write, rpmsg_input to send the request parameters to rpmsg_rpc_client_send, exactly as the code snippet shared with poll to wait for answer.

**Master application(linux_rpc_demod.c)**

- Polls for request from slave until a termination request is received

arnopo reviewed (changes requested):
Hello @viswaharsha

i did not check all in details as the split between amplication and service layer as to be reworked  from my point of view.

As already mentionned the rpc service layer should not implement the rpc function (e.g. rpmsg_handle_open) but should only by a layer that package the rpc requests and answers.
Every object related to your  OPEN/CLOSE/WRITE/READ/ACK_STATUS/TERM/INPUT should be  handled in the application.

Please also:
- split your patchset in at least 2 commits one for ce service one for the sample
- have  a look to the "open-amp lib Continuous Integration" that report error on compilation

viswaharsha commented:
Hi @arnopo,
Thanks for the comments, but still I feel it would be more clarified if I have a call with you regarding this commit for further process. Could you kindly share your contact details to my mail cviswa.harsha_ext@ltts.com to schedule a call whenever you are available?

viswaharsha commented on lib/service/rpmsg/rpc/rpmsg_rpc_client.c:
```diff
@@ -0,0 +1,132 @@
+				if (len > rpc->respbuf_len)
+					len = rpc->respbuf_len;
+				rpc->respbuf = data;
+			}
```
Hi,
I have included a polling mechanism in application defined functions. These functions before sending request, sets up flag `(void)atomic_flag_test_and_set(&rpc->nacked)` and polls until this flag is cleared. This flag is cleared in the endpoint callback function on service layer `atomic_flag_clear(&rpc->nacked)`, once the response/answer is received. Through this the user will be able to define functions asynchronous or synchronous by setting flag.

If there is a callback from endpoint callback in service layer to user defined functions in application layer, the callback can only be defined either asynchronous or synchronous, also abstraction breaks between application layer and service layer.

viswaharsha commented on lib/service/rpmsg/rpc/rpmsg_rpc_server.c:
```diff
@@ -0,0 +1,285 @@
+	return ret > 0 ?  0 : ret;
+}
+
+int linux_rpmsg_rpc_server_init(int id, void *data,
```
I have created a hash table taking structure that maps id(`OPEN_ID`) with its defined callback function(eg.`rpmsg_handle_open`), as accessing the mapped callback function takes relatively less time complexity during endpoint callback.

arnopo commented on lib/service/rpmsg/rpc/rpmsg_rpc_server.c:
```diff
@@ -0,0 +1,129 @@
+
+struct rpmsg_rpc_services *find_service(int id)
+{
+	int hash_index = id % TABLE_SIZE;
```
I don't understand this function 
rpc id should be independent from the index of the table.
That said this function should be probably reworked based on previous comment concerning the management of the rpmsg_rpc_services table

arnopo commented on lib/service/rpmsg/rpc/rpmsg_rpc_server.c:
```diff
@@ -0,0 +1,129 @@
+{
+	unsigned char buf[MAX_BUF_LEN];
+	unsigned long int id;
+
```
Suppress blank line

arnopo commented on lib/service/rpmsg/rpc/rpmsg_rpc_server.c:
```diff
@@ -0,0 +1,129 @@
+int err_cnt;
+static struct rpmsg_rpc_services *table[TABLE_SIZE];
+
+void *copy_from_shbuf(void *dst, void *shbuf, int len)
```
static void *copy_from_shbuf(void *dst, void *shbuf, int len)

arnopo commented on lib/service/rpmsg/rpc/rpmsg_rpc_server.c:
```diff
@@ -0,0 +1,129 @@
+	 * E.g. For now, we only use UIO device memory in Linux.
+	 */
+	if (len > MAX_BUF_LEN)
+		len = MAX_BUF_LEN;
```
if len > MAX_BUF_LEN , it will probably not work, just return an error

arnopo commented on lib/service/rpmsg/rpc/rpmsg_rpc_server.c:
```diff
@@ -0,0 +1,129 @@
+	return NULL;
+}
+
+int linux_rpmsg_endpoint_server_cb(struct rpmsg_endpoint *ept, void *data,
```
Why "linux"?

arnopo commented on lib/service/rpmsg/rpc/rpmsg_rpc_server.c:
```diff
@@ -0,0 +1,129 @@
+		len = MAX_BUF_LEN;
+	copy_from_shbuf(buf, data, len);
+	id = *buf;
+	struct rpmsg_rpc_services *service = find_service(id);
```
service variable declaration to move to the begining of the function

arnopo commented on lib/service/rpmsg/rpc/rpmsg_rpc_server.c:
```diff
@@ -0,0 +1,129 @@
+	} else {
+		LPRINTF("\nHandling remote procedure call errors:\r\n");
+		raw_printf("rpc id %ld\r\n", id);
+		err_cnt++;
```
Suppress the err_cnt and concatenate the  raw_printf and LPRINTF print in one metal_log(METAL_LOG_ERROR,...)

arnopo reviewed (commented):
Thanks for thze update. The split between service and application seems  to be close to be ok.
Main remaining points:
- rpmsg_rpc_services table to rework to define in application and avoid malloc  in service that could be not supported on some platforms
-  move the atomic management (rpc->naked) in application by replacing it by a response callback.

I haven't had time to look at the application part in depth yet.  But i propose to focus first on the service part. The appliaction part has less contraint as dedicated for the Linux implementation.

arnopo commented on lib/service/rpmsg/rpc/rpmsg_rpc_server.c:
```diff
@@ -0,0 +1,104 @@
+
+	if (len > MAX_BUF_LEN)
+		return -EINVAL;
+	copy_from_shbuf(buf, data, len);
```
no clear for me what you want to do here witjh thre function copy_from_shbuf
why not just define a rpc_msg struct 
```
struct rpmsg_rpc {
	unsigned int id;
	size_t param_size; /* optional can be deduced from the rpmsg size */
        uint8_t param[0] 
};

```
and then you have just to cast the message 

```
	struct rpmsg_rpc *msg = (struct rpmsg_rpc *) data;

	service = find_service(msg->id);
	if (service) {
		service->respcb.cb_function(msg->param, msg->param_size, ept);

```

arnopo commented on lib/include/openamp/rpmsg_rpc_client_server.h:
```diff
@@ -0,0 +1,198 @@
+ * @ept_addr: address of the endpoint used by RPC
+ * @ept_raddr: remote address of the endpoint used by RPC
+ * @poll_arg: pointer to poll function argument
+ * @poll: poll function
```
poll field seems to me uselless, can be full handled in application part. can be also removed from rpmsg_rpc_data struct

arnopo commented on lib/service/rpmsg/rpc/rpmsg_rpc_client.c:
```diff
@@ -0,0 +1,126 @@
+int rpmsg_rpc_client_send(struct rpmsg_rpc_data *rpc,
+			  int rp, void *request_param,
+			  size_t req_param_size, void *response_param,
+			  size_t resp_param_size, void *cb)
```
Here the APi inpose that you serialize your RPC request. If you call 2 time this function you will overwrite some rpc structure fileds.
i would suggest you a similar approche on client thatn the serveur using` rpmsg_rpc_services` struct.
at ini you provide the table and when an answer is received, you call the CB depending on the ID.  This would allow to parallelize the requests.

arnopo commented on lib/service/rpmsg/rpc/rpmsg_rpc_client.c:
```diff
@@ -0,0 +1,126 @@
+	if (!rpc)
+		return -EINVAL;
+
+	metal_spinlock_acquire(&rpc->buflock);
```
this seems to protect nothing from overwrite for instance if the function is recalled  before rpmsg_endpoint_client_cb is received this not prevent overwrite

arnopo commented on lib/include/openamp/rpmsg_rpc_client_server.h:
```diff
@@ -0,0 +1,198 @@
+	rpmsg_rpc_poll poll;
+	void *poll_arg;
+	rpmsg_rpc_shutdown_cb shutdown_cb;
+	metal_mutex_t lock;
```
seems not to be used anymore

arnopo commented on lib/include/openamp/rpmsg_rpc_client_server.h:
```diff
@@ -0,0 +1,198 @@
+ */
+struct rpmsg_rpc_data {
+	struct rpmsg_endpoint ept;
+	atomic_int nacked;
```
used only in application. to remove from the srtucture and define it in application

arnopo commented on lib/service/rpmsg/rpc/rpmsg_rpc_client.c:
```diff
@@ -0,0 +1,143 @@
+	if (!rpc)
+		return -EINVAL;
+
+	rpc->respbuf = response_param;
```
you send a pointer to the remote side that has only a local reference .

arnopo reviewed (changes requested):
Here is few new remarks,
To accelerate the review and give you a better view of what  i have in mind i will send you by email some fixes on top of your pull request.

viswaharsha commented:
Hi @arnopo ,

Getting some cmake version error in the Continuous Integration Check

arnopo commented:
> Hi @arnopo ,
> 
> Getting some cmake version error in the Continuous Integration Check

I send a PR to fix the CI issue: https://github.com/OpenAMP/open-amp/pull/299. When  integrated you should be able to rebase to rerun the CI. But this does not bock the code review.

@edmooring: could you have a look to this PR please?

arnopo commented on lib/include/openamp/rpmsg_rpc_client_server.h:
```diff
@@ -0,0 +1,246 @@
+ *
+ * This function is to initialize the remote procedure call
+ * global data. RPMsg RPC will send request to remote and
+ * wait for callback and load services to table
```
comment to update

arnopo commented on lib/include/openamp/rpmsg_rpc_client_server.h:
```diff
@@ -0,0 +1,246 @@
+ *
+ * return 0 for success
+ */
+int rpmsg_endpoint_client_cb(struct rpmsg_endpoint *ept,
```
this function does not need to be exported in the .h

arnopo commented on lib/service/rpmsg/rpc/rpmsg_rpc_client.c:
```diff
@@ -0,0 +1,117 @@
+
+}
+
+int rpmsg_endpoint_client_cb(struct rpmsg_endpoint *ept,
```
static function

arnopo commented on lib/include/openamp/rpmsg_rpc_client_server.h:
```diff
@@ -0,0 +1,246 @@
+ *
+ * return 0 for success
+ */
+int rpmsg_endpoint_server_cb(struct rpmsg_endpoint *ept, void *data,
```
Same for this function

arnopo commented on lib/include/openamp/rpmsg_rpc_client_server.h:
```diff
@@ -0,0 +1,246 @@
+ * This function sends RPC request
+ *
+ * @rpcs: pointer to server rpc struct
+ * @rp:  function id
```
For readablity, could you  rename this param rpc_id

arnopo commented on lib/service/rpmsg/rpc/rpmsg_rpc_client.c:
```diff
@@ -0,0 +1,117 @@
+#include <fcntl.h>
+
+#define LPRINTF(format, ...) printf(format, ##__VA_ARGS__)
+#define LPERROR(format, ...) LPRINTF("ERROR: " format, ##__VA_ARGS__)
```
these macro seems not used , you can clean them

arnopo commented on lib/service/rpmsg/rpc/rpmsg_rpc_client.c:
```diff
@@ -0,0 +1,117 @@
+	struct rpmsg_rpc_data *rpc;
+
+	rpc = metal_container_of(ept, struct rpmsg_rpc_data, ept);
+	(void)ept;
```
put this line after 
`struct rpmsg_rpc_data *rpc;`

arnopo commented on lib/service/rpmsg/rpc/rpmsg_rpc_server.c:
```diff
@@ -0,0 +1,95 @@
+		service->respcb.cb_function(buf, rpcs);
+	} else {
+		LPRINTF("Handling remote procedure call errors: rpc id %ld\r\n",
+			id);
```
use LPERROR

arnopo commented on lib/include/openamp/rpmsg_rpc_client_server.h:
```diff
@@ -0,0 +1,246 @@
+			  rpmsg_rpc_shutdown_cb shutdown_cb,
+			  const struct rpmsg_rpc_client_services *services,
+			  int len);
+
```
Sorry just notice that now...
The name service and endpoint address should not be part of the rpmsg_rpc_client_init and rpmsg_rpc_server_init param. a sepecific APi will be added if someone need to specify them.
```
int rpmsg_rpc_client_init(struct rpmsg_rpc_data *rpc,
			  struct rpmsg_device *rdev,
			  rpmsg_rpc_shutdown_cb shutdown_cb,
			  const struct rpmsg_rpc_client_services *services,
			  int len);
```
i describe impact in the function body below

arnopo commented on lib/service/rpmsg/rpc/rpmsg_rpc_client.c:
```diff
@@ -0,0 +1,117 @@
+	ret = rpmsg_create_ept(&rpc->ept, rdev,
+			       ept_name, ept_addr, ept_raddr,
+			       rpmsg_endpoint_client_cb,
+			       rpmsg_service_client_unbind);
```
replace by
```
ret = rpmsg_create_ept(&rpc->ept, rdev,
			       RPMSG_RPC_SERVICE_NAME , RPMSG_ADDR_ANY, RPMSG_ADDR_ANY,
			       rpmsg_endpoint_client_cb,
			       rpmsg_service_client_unbind);

```	
to suppress related function params

arnopo commented on apps/examples/linux_rpc_demo/linux-rpmsg-rpc-demo.h:
```diff
@@ -0,0 +1,77 @@
+#ifndef RPMSG_RPC_DEMO_H
+#define RPMSG_RPC_DEMO_H
+
+#define RPMSG_SERVICE_NAME         "rpmsg-openamp-demo-channel"
```
The name service should be defined inrpmsg_rpc_client_server.h with following define

#define RPMSG_RPC_SERVICE_NAME "rpmsg-rpc"

arnopo reviewed (changes requested):
Tested in my linux and it work pretty well !
to help to reproduce the example, please could you add a short README file in app/exemples/linux_rpc_demo, that explains how to run the example?

then some remaining cleanup and optimization ( see my comments), 
but it is  close to be good from my point of view to be integrated

Thanks,
Arnaud

arnopo commented on lib/service/rpmsg/rpc/rpmsg_rpc_server.c:
```diff
@@ -0,0 +1,95 @@
+	service = find_service(rpcs, id);
+
+	if (service) {
+		service->respcb.cb_function(buf, rpcs);
```
The callback returns a status. Status should be tested and an error message printed

edmooring commented on apps/examples/linux_rpc_demo/linux_rpc_demo.c:
```diff
@@ -0,0 +1,599 @@
+static struct rpmsg_rpc_clt *rpmsg_default_rpc;
+static int fd, bytes_written, bytes_read;
+static struct polling poll;
+static atomic_int nacked;
```
This name is misleading. "NACK" is "negative acknowledgment", saying "no" to something. Here it's a flag saying that the other end has processed the request.

edmooring reviewed (approved):
Neither of my comments are show stoppers. OK to go as is. Prettying things up can come in another patch.

arnopo commented:
@viswaharsha :
Could you take into account @edmooring comment, before we merge?
Thanks
arnaud

viswaharsha commented:
> @viswaharsha :
> Could you take into account @edmooring comment, before we merge?
> Thanks
> arnaud

Yeah sure, @arnopo, @edmooring .

arnopo commented:
> > @viswaharsha :
> > Could you take into account @edmooring comment, before we merge?
> > Thanks
> > arnaud
> 
> Yeah sure, @arnopo, @edmooring .

Hi @viswaharsha,
Do you think you will find time to update the PR, so  we integratae your work in the october release?
Thanks!
Arnaud

viswaharsha commented:
Hi @arnopo , I was tied up with other works, so was not able to focus on this PR. I'll send you the updated PR by this weekend.
Also, can you please elaborate @edmooring comment regarding OPENAMP_PACKED_BEGIN/END?

arnopo commented:
> Hi @arnopo , I was tied up with other works, so was not able to focus on this PR. I'll send you the updated PR by this weekend.
> Also, can you please elaborate @edmooring comment regarding OPENAMP_PACKED_BEGIN/END?

just encapsulate the structure with the macro ( refer to [vring_desc struct](https://github.com/OpenAMP/open-amp/blob/master/lib/include/openamp/virtio_ring.h#L39l) as example)
to understand packed attribute use : https://sites.google.com/site/eganya/Home/links/gcc---packed-structures

> Also taking his second comment can we change "nacked" as "acked" ?

As you mentioned in code it is used to : " wait for response" so something such as "wait_resp" might be more consistent with its use.

viswaharsha commented:
Hi @arnopo @edmooring ,
Done with the requested changes.
Requesting your review.
Thanks!
**Viswa Harsha C**

arnopo commented on lib/include/openamp/rpmsg_rpc_client_server.h:
```diff
@@ -0,0 +1,208 @@
+#endif
+
+#define OPENAMP_PACKED_BEGIN
+#define OPENAMP_PACKED_END __attribute__((__packed__))
```
Do not define new macro  but reuse exiting one.

Add `#include <metal/compiler.h>` and reuse  `METAL_PACKED_BEGIN`/` METAL_PACKED_END`; as done in https://github.com/OpenAMP/open-amp/blob/master/lib/include/openamp/virtio_ring.h#L39l
the METAL_PACKED_xx macro depend on the compiler

arnopo commented on lib/include/openamp/rpmsg_rpc_client_server.h:
```diff
@@ -0,0 +1,208 @@
+ * @params: answer params
+ *
+ */
+OPENAMP_PACKED_BEGIN
```
METAL_PACKED_BEGIN

arnopo commented on lib/include/openamp/rpmsg_rpc_client_server.h:
```diff
@@ -0,0 +1,208 @@
+	uint32_t id;
+	int32_t status;
+	unsigned char params[MAX_BUF_LEN];
+} OPENAMP_PACKED_END;
```
METAL_PACKED_END

viswaharsha commented:
Hi @arnopo , kindly review.

arnopo commented:
Thanks @viswaharsha  for your work it will be part of next release!
